src/WebSockets.jl

```diff
@@ -94,22 +96,27 @@ function write_fragment(io::IO, islast::Bool, data::Array{UInt8}, opcode)
   # 2. set a mask
   # 3. xor data with mask
 
-  if l <= 125
-    write(io, b1)
-    write(io, @compat UInt8(l))
-    write(io, data)
-  elseif l <= typemax(UInt16)
-    write(io, b1)
-    write(io, @compat UInt8(126))
-    write(io, hton(@compat UInt16(l)))
-    write(io, data)
-  elseif l <= typemax(UInt64)
-    write(io, b1)
-    write(io, @compat UInt8(127))
-    write(io, hton(@compat UInt64(l)))
-    write(io, data)
-  else
-    error("Attempted to send too much data for one websocket fragment\n")
+  isa(io, TCPSock) && lock(io.lock)
+  try
+    if l <= 125
+      write(io, b1)
+      write(io, @compat UInt8(l))
+      write(io, data)
+    elseif l <= typemax(UInt16)
+      write(io, b1)
+      write(io, @compat UInt8(126))
+      write(io, hton(@compat UInt16(l)))
+      write(io, data)
+    elseif l <= typemax(UInt64)
+      write(io, b1)
+      write(io, @compat UInt8(127))
+      write(io, hton(@compat UInt64(l)))
+      write(io, data)
+    else
+      error("Attempted to send too much data for one websocket fragment\n")
+    end
+  finally
```

Comment:
You can optimize the multiple writes above by calling `Base.buffer_writes` once on the socket at construction time (e.g. https://github.com/JuliaLang/julia/blob/3f8d5a9aabfc76e445ce5d58875b4327c525454e/base/multi.jl#L145) and then finally flushing once after all individual writes of a fragment - e.g. https://github.com/JuliaLang/julia/blob/3f8d5a9aabfc76e445ce5d58875b4327c525454e/base/multi.jl#L227

This results in the TCPSocket aggregating individual writes into a local buffer (upto 65K) and actually writing to the socket upon a `flush`. Larger than 65K writes get written immediately.